File: 617.合并二叉树.java
import java.util.Stack;

/*
 * @lc app=leetcode.cn id=617 lang=java
 *
 * [617] 合并二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode mergeTrees(TreeNode root1, TreeNode root2) {
        if (root1 == null && root2 == null) {
            return null;
        }
        if (root1 == null) return root2;
        if (root2 == null) return root1;

        Stack<TreeNode> stack = new Stack<>();
        stack.push(root1);
        stack.push(root2);

        while (!stack.isEmpty()) {
            TreeNode cur2 = stack.pop();
            TreeNode cur1 = stack.pop();

            cur1.val += cur2.val;

            if (cur1.left != null && cur2.left != null){
                stack.push(cur1.left);
                stack.push(cur2.left);
            } else {
                if (cur1.left == null) {
                    cur1.left = cur2.left;
                }
            }
            if (cur1.right != null && cur2.right != null){
                stack.push(cur1.right);
                stack.push(cur2.right);
            } else {
                if (cur1.right == null) {
                    cur1.right = cur2.right;
                }
            }
        }
        return root1;

    }
}
// @lc code=end

File: 617.合并二叉树.py
#
# @lc app=leetcode.cn id=617 lang=python
#
# [617] 合并二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def mergeTrees1(self, root1, root2):
        """
        :type root1: TreeNode
        :type root2: TreeNode
        :rtype: TreeNode
        """
        if not root1 and not root2:
            return None
        if not root1:
            return root2
        if not root2:
            return root1
        
        root = TreeNode(root1.val + root2.val)
        root.left = self.mergeTrees(root1.left, root2.left)
        root.right = self.mergeTrees(root1.right, root2.right)

        return root
# @lc code=end

